### PR TITLE
Improve CLI 

### PR DESCRIPTION
**Pull request by OmranAbdelkafi**
## Description 
Enhance help menu and command line options
Update encoder user guide to match the help menu 
## Related issues
#779
#945
## Performance 
No impact

### CODE REVIEW

**Comment by hassount** [comment]
@OmranAbdelkafi can you please add the reference to the issues that this PR addresses?

**Comment by kylophone** [comment]
> Update encoder user guide to match the help menu

Will this diverge again? I wonder if there's a way to templatize this.

File: Source/App/EncApp/EbAppConfig.c [hunk 1]
```diff
@@ -110,7 +113,6 @@
 #define HME_LEVEL2_WIDTH "-hme-l2-w"
 #define HME_LEVEL2_HEIGHT "-hme-l2-h"
 #define SCREEN_CONTENT_TOKEN "-scm"
-#define INTRABC_MODE_TOKEN "-intrabc-mode"
```
Comment (adaria) [inline Source/App/EncApp/EbAppConfig.c]: @OmranAbdelkafi was there an issue with -intrabc-mode or has it just slipped through when merged via the cs2 branch? (It was merged into master but not cs2). I can look at it if there were issues, let me know. Thanks!

**Comment by adaria** [comment]
@OmranAbdelkafi @mazharaIntel How does this PR relate to the getopt effort under way in the [use_getopt](https://github.com/OpenVisualCloud/SVT-AV1/tree/use_getopt) branch? Is it a fix while we wait for getopt or we decided not to go with getopt for now?

**Comment by OmranAbdelkafi** [comment]
> This PR is incomplete, e.g. -altref-strength and -altref-nframes are still single dash in AppConfig.c although they're described as --altref-strength and --altref-nframes in the encoder doc. Are there more changes that have not been ported from cs2?
> 
> --help and -help work but not -h which should be added if we go with -- (double dash) cli options.
> 
> Also, moving to double dash in the encoder CLI would make it inconsistent with the decoder CLI which has single-dash options.
> 
> I recommend not merging this PR into master.


all long tokens( more than a single letter ) were taken into consideration as we added the new ones with double dash while the old one and in order to avoid duplication we just adapted them to the double dash format without rewriting them. For the "-h" case all encoders use a single dash when it comes to short token and they have their equivalent using double dash which is "--height" . We kept " -help " exclusively so we wont break the build tests.
Concerning the decoder I think it is better that we adapt the decoder tokens to the new format since they are not that many rather than just not adopting the new encoder cli format

**Comment by OmranAbdelkafi** [comment]
> @OmranAbdelkafi @mazharaIntel How does this PR relate to the getopt effort under way in the [use_getopt](https://github.com/OpenVisualCloud/SVT-AV1/tree/use_getopt) branch? Is it a fix while we wait for getopt or we decided not to go with getopt for now?

The getopt has a lot of dependency issues that we are still working on , that s why we are using this PR instead

**Comment by adaria** [comment]
Got it. My bad re -h option, of course it's height [face-palm]
I've rebased the branch and put -intrabc-mode back

However, there are still many tokens which are not moved to double-dash, have they not been ported from cs2 or you are working on them? I tried to catch the renamed tokens but may have missed a few:

- [ ] CHANNEL_NUMBER_TOKEN -nch
- [ ] ERROR_FILE_TOKEN -errlog
- [ ] QP_FILE_TOKEN -qp-file
- [ ] INPUT_STAT_FILE_TOKEN -input-stat-file
- [ ] OUTPUT_STAT_FILE_TOKEN -output-stat-file
- [ ] STAT_FILE_TOKEN -stat-file
- [ ] INPUT_PREDSTRUCT_FILE_TOKEN -pred-struct-file
- [ ] BUFFERED_INPUT_TOKEN -nb
- [ ] BASE_LAYER_SWITCH_MODE_TOKEN -base-layer-switch-mode
- [ ] USE_QP_FILE_TOKEN -use-q-file
- [ ] FRAME_RATE_TOKEN -fps
- [ ] FRAME_RATE_NUMERATOR_TOKEN -fps-num
- [ ] FRAME_RATE_DENOMINATOR_TOKEN -fps-denom
- [ ] ENCMODE2P_TOKEN -enc-mode-2p
- [ ] HIERARCHICAL_LEVELS_TOKEN -hierarchical-levels
- [ ] PRED_STRUCT_TOKEN -pred-struct
- [ ] PROFILE_TOKEN -profile
- [ ] TIER_TOKEN -tier
- [ ] LEVEL_TOKEN -level
- [ ] FILM_GRAIN_TOKEN -film-grain
- [ ] INTRA_REFRESH_TYPE_TOKEN -irefresh-type
- [ ] LOOP_FILTER_DISABLE_TOKEN -dlf
- [ ] CDEF_MODE_TOKEN -cdef-mode
- [ ] SG_FILTER_MODE_TOKEN -sg-filter-mode
- [ ] WN_FILTER_MODE_TOKEN -wn-filter-mode
- [ ] SUBPEL_TOKEN -subpel
- [ ] OBMC_TOKEN -obmc
- [ ] PRED_ME_TOKEN -pred-me
- [ ] BIPRED_3x3_TOKEN -bipred-3x3
- [ ] COMPOUND_LEVEL_TOKEN -compound
- [ ] PIC_BASED_RATE_EST_TOKEN -pic-based-rate-est
- [ ] USE_DEFAULT_ME_HME_TOKEN -use-default-me-hme
- [ ] HME_ENABLE_TOKEN -hme
- [ ] HME_L0_ENABLE_TOKEN -hme-l0
- [ ] HME_L1_ENABLE_TOKEN -hme-l1
- [ ] HME_L2_ENABLE_TOKEN -hme-l2
- [ ] SEARCH_AREA_WIDTH_TOKEN -search-w
- [ ] SEARCH_AREA_HEIGHT_TOKEN -search-h
- [ ] NUM_HME_SEARCH_WIDTH_TOKEN -num-hme-w
- [ ] NUM_HME_SEARCH_HEIGHT_TOKEN -num-hme-h
- [ ] HME_SRCH_T_L0_WIDTH_TOKEN -hme-tot-l0-w
- [ ] HME_SRCH_T_L0_HEIGHT_TOKEN -hme-tot-l0-h
- [ ] SCREEN_CONTENT_TOKEN -scm
- [ ] INTRABC_MODE_TOKEN -intrabc-mode
- [ ] HBD_MD_ENABLE_TOKEN -hbd-md
- [ ] PALETTE_TOKEN -palette
- [ ] OLPD_REFINEMENT_TOKEN -olpd-refinement
- [ ] RATE_CONTROL_ENABLE_TOKEN -rc
- [ ] TARGET_BIT_RATE_TOKEN -tbr
- [ ] MAX_QP_TOKEN -max-qp
- [ ] VBV_BUFSIZE_TOKEN -vbv-bufsize
- [ ] MIN_QP_TOKEN -min-qp
- [ ] ADAPTIVE_QP_ENABLE_TOKEN -adaptive-quantization
- [ ] LOOK_AHEAD_DIST_TOKEN -lad
- [ ] SUPER_BLOCK_SIZE_TOKEN -sb-size
- [ ] TILE_ROW_TOKEN -tile-rows
- [ ] TILE_COL_TOKEN -tile-columns
- [ ] SQ_WEIGHT_TOKEN -sqw
- [ ] CHROMA_MODE_TOKEN -chroma-mode
- [ ] DISABLE_CFL_TOKEN -dcfl
- [ ] SCENE_CHANGE_DETECTION_TOKEN -scd
- [ ] INJECTOR_TOKEN -inj
- [ ] INJECTOR_FRAMERATE_TOKEN -inj-frm-rt
- [ ] SPEED_CONTROL_TOKEN -speed-ctrl
- [ ] ASM_TYPE_TOKEN -asm
- [ ] UNPIN_LP1_TOKEN -unpin-lp1

**Comment by 1480c1** [comment]
Although this is a bit late, might I suggest doing something similar to ffmpeg's `-filter` option for the optimization related opts with a generic flag like `-opts` or something that takes a string that essentially is the config file options, but instead of delimited by `\n` then `:`, it's delimited by `:` then `=`, rather than keeping the massive amount of kebab flags we currently have? (https://github.com/FFmpeg/FFmpeg/blob/master/fftools/cmdutils.c for what ffmpeg is currently doing)

A different suggestion is separate out the commonly used flags and the optimization related flags in the docs with the commonly used ones on the top and the rest at the bottom and use a format that can be consumed and used to produce a man page like https://github.com/rtomayko/ronn or pandoc

**Comment by OmranAbdelkafi** [comment]
> Got it. My bad re -h option, of course it's height [face-palm]
> I've rebased the branch and put -intrabc-mode back
> 
> However, there are still many tokens which are not moved to double-dash, have they not been ported from cs2 or you are working on them? I tried to catch the renamed tokens but may have missed a few:
> 
> * [ ]  CHANNEL_NUMBER_TOKEN -nch
> * [ ]  ERROR_FILE_TOKEN -errlog
> * [ ]  QP_FILE_TOKEN -qp-file
> * [ ]  INPUT_STAT_FILE_TOKEN -input-stat-file
> * [ ]  OUTPUT_STAT_FILE_TOKEN -output-stat-file
> * [ ]  STAT_FILE_TOKEN -stat-file
> * [ ]  INPUT_PREDSTRUCT_FILE_TOKEN -pred-struct-file
> * [ ]  BUFFERED_INPUT_TOKEN -nb
> * [ ]  BASE_LAYER_SWITCH_MODE_TOKEN -base-layer-switch-mode
> * [ ]  USE_QP_FILE_TOKEN -use-q-file
> * [ ]  FRAME_RATE_TOKEN -fps
> * [ ]  FRAME_RATE_NUMERATOR_TOKEN -fps-num
> * [ ]  FRAME_RATE_DENOMINATOR_TOKEN -fps-denom
> * [ ]  ENCMODE2P_TOKEN -enc-mode-2p
> * [ ]  HIERARCHICAL_LEVELS_TOKEN -hierarchical-levels
> * [ ]  PRED_STRUCT_TOKEN -pred-struct
> * [ ]  PROFILE_TOKEN -profile
> * [ ]  TIER_TOKEN -tier
> * [ ]  LEVEL_TOKEN -level
> * [ ]  FILM_GRAIN_TOKEN -film-grain
> * [ ]  INTRA_REFRESH_TYPE_TOKEN -irefresh-type
> * [ ]  LOOP_FILTER_DISABLE_TOKEN -dlf
> * [ ]  CDEF_MODE_TOKEN -cdef-mode
> * [ ]  SG_FILTER_MODE_TOKEN -sg-filter-mode
> * [ ]  WN_FILTER_MODE_TOKEN -wn-filter-mode
> * [ ]  SUBPEL_TOKEN -subpel
> * [ ]  OBMC_TOKEN -obmc
> * [ ]  PRED_ME_TOKEN -pred-me
> * [ ]  BIPRED_3x3_TOKEN -bipred-3x3
> * [ ]  COMPOUND_LEVEL_TOKEN -compound
> * [ ]  PIC_BASED_RATE_EST_TOKEN -pic-based-rate-est
> * [ ]  USE_DEFAULT_ME_HME_TOKEN -use-default-me-hme
> * [ ]  HME_ENABLE_TOKEN -hme
> * [ ]  HME_L0_ENABLE_TOKEN -hme-l0
> * [ ]  HME_L1_ENABLE_TOKEN -hme-l1
> * [ ]  HME_L2_ENABLE_TOKEN -hme-l2
> * [ ]  SEARCH_AREA_WIDTH_TOKEN -search-w
> * [ ]  SEARCH_AREA_HEIGHT_TOKEN -search-h
> * [ ]  NUM_HME_SEARCH_WIDTH_TOKEN -num-hme-w
> * [ ]  NUM_HME_SEARCH_HEIGHT_TOKEN -num-hme-h
> * [ ]  HME_SRCH_T_L0_WIDTH_TOKEN -hme-tot-l0-w
> * [ ]  HME_SRCH_T_L0_HEIGHT_TOKEN -hme-tot-l0-h
> * [ ]  SCREEN_CONTENT_TOKEN -scm
> * [ ]  INTRABC_MODE_TOKEN -intrabc-mode
> * [ ]  HBD_MD_ENABLE_TOKEN -hbd-md
> * [ ]  PALETTE_TOKEN -palette
> * [ ]  OLPD_REFINEMENT_TOKEN -olpd-refinement
> * [ ]  RATE_CONTROL_ENABLE_TOKEN -rc
> * [ ]  TARGET_BIT_RATE_TOKEN -tbr
> * [ ]  MAX_QP_TOKEN -max-qp
> * [ ]  VBV_BUFSIZE_TOKEN -vbv-bufsize
> * [ ]  MIN_QP_TOKEN -min-qp
> * [ ]  ADAPTIVE_QP_ENABLE_TOKEN -adaptive-quantization
> * [ ]  LOOK_AHEAD_DIST_TOKEN -lad
> * [ ]  SUPER_BLOCK_SIZE_TOKEN -sb-size
> * [ ]  TILE_ROW_TOKEN -tile-rows
> * [ ]  TILE_COL_TOKEN -tile-columns
> * [ ]  SQ_WEIGHT_TOKEN -sqw
> * [ ]  CHROMA_MODE_TOKEN -chroma-mode
> * [ ]  DISABLE_CFL_TOKEN -dcfl
> * [ ]  SCENE_CHANGE_DETECTION_TOKEN -scd
> * [ ]  INJECTOR_TOKEN -inj
> * [ ]  INJECTOR_FRAMERATE_TOKEN -inj-frm-rt
> * [ ]  SPEED_CONTROL_TOKEN -speed-ctrl
> * [ ]  ASM_TYPE_TOKEN -asm
> * [ ]  UNPIN_LP1_TOKEN -unpin-lp1

Like I mentioned in my previous comment the tokens that we did not use to have were introduced to the code with double dash (like "--frames" option) , also for the tokens that got name changes we duplicated new equivalent tokens for that( for example --enable-restoration-filtering token is the new equivalent token for "-restoration-filtering"). In the other hand for the old tokens that we didn't change at all(No renaming no new equivalent token just like -hierarchical-levels option) we kept them the same way they are defined within the code and we just changed the way they are handled in the command line so they could be recognized when there are written in a double dash format and that was done so we could avoid duplication and save some lines. 
And also we have to keep in mind that we are still supporting the old tokens format ( single dash option whether short or long) so we don't break any actual testing/scripts still using such format.
In other words , you can try : 
case1 : SvtAV1EncApp.exe -i sample.yuv -w 640-h 360 -fps 60 -intra-period 32 -enc-mode 8 -q 20
cas2 : SvtAV1EncApp.exe --input sample.yuv --width 640 --height 360 --fps 60 --keyint 32 --preset 8 --qp 20
=> Both cases would work and give you the exact same bitstreams , but case 1 would show you a warning saying that you are using old tokens and they might be deprecated soon(-fps,-intra-period,-enc-mode) and will point you to the equivalent new ones (--fps , --keyint , --preset)
I hope that clarify the issue

**Comment by adaria** [comment]
I does clarify, thank you @OmranAbdelkafi ! I missed the treatment of single dash as double dash in code. I therefore updated to --intrabc-mode in the user guide so it's consistent with other options.
